From eb742581df5b7c4372fa566c7af3be98d94ae7bc Mon Sep 17 00:00:00 2001
From: Minghuan Lian <Minghuan.Lian@freescale.com>
Date: Fri, 3 Jul 2015 14:57:07 +0800
Subject: [PATCH 048/124] iommu/arm-smmu: Add support layerscape PCIe device
 stream ID

Some Layerscape PCIe devices use stream ID instead of the physical
device ID. The patch stores this stream ID to dma_alias_devid.
We should support PCIe bridges, because PCIe controller also needs
MSI interrupt.

Signed-off-by: Minghuan Lian <Minghuan.Lian@freescale.com>
[Stuart: manual merge of patch]
Signed-off-by: Stuart Yoder <stuart.yoder@freescale.com>
---
 drivers/iommu/arm-smmu.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/iommu/arm-smmu.c b/drivers/iommu/arm-smmu.c
index 40d3cb9..5da0521 100644
--- a/drivers/iommu/arm-smmu.c
+++ b/drivers/iommu/arm-smmu.c
@@ -1389,9 +1389,6 @@ static int arm_smmu_add_pci_device(struct pci_dev *pdev)
 	if (IS_ERR(group))
 		return PTR_ERR(group);
 
-	if (pci_is_bridge(pdev))
-		goto out_put_group;
-
 	cfg = iommu_group_get_iommudata(group);
 	if (!cfg) {
 		cfg = kzalloc(sizeof(*cfg), GFP_KERNEL);
@@ -1430,6 +1427,9 @@ static int arm_smmu_add_pci_device(struct pci_dev *pdev)
 	}
 	cfg->streamids[0] = streamid;
 	cfg->mask = 0x7c00;
+
+	pdev->dev_flags |= PCI_DEV_FLAGS_DMA_ALIAS_DEVID;
+	pdev->dma_alias_devid = streamid;
 #endif
 
 	return 0;
-- 
2.1.0.27.g96db324

